util/network-devp2p/src/discovery.rs

```diff
@@ -169,6 +169,9 @@ pub struct Discovery<'a> {
 	discovery_nodes: HashSet<NodeId>,
 	node_buckets: Vec<NodeBucket>,
 
+	// these nodes can't be used for syncing.
+	bootnodes: HashSet<NodeId>,
```

Comment:
I still don't understand why do we need this? Those nodes doesn't seem to be added to `NodeBucket`s so they never going to be returned by `FIND_NODES` query.

So it seems that we are solving the problem here, but only partially by accepting `PING`s from nodes without TCP port, but not handling them properly as part of the discoverable network.

Reply:
Also note that this collection is unbounded, so we will slowly leak memory here. It's exploitable by sending a bunch of PINGs from random ports.

Reply:
@tomusdrw the node will be added to `NodeBucket`

https://github.com/paritytech/parity-ethereum/pull/10705/files/783a66df3bf69b52f4372fcd32efc7f538451b6b#diff-3193558172eb0ec141a7afd53b7a1118R272

we're simply preventing the node from being added to the `Host::nodes` which is used for syncing.

and yes, the set is unbounded. wanted to open discussion on the PR on what conditions it should be cleared.

Reply:
I see, but I don't get why we need a separate set for this, why can't we have a `has_tcp_port()` method on the `entry.endpoint` itself similar to how we have `is_valid` now?

Reply:
Maybe we could also rename the methods to `is_valid_sync_node()` & `is_valid_discovery_node()` for clarity.